TFLu: Update builds with CMSIS-NN

mansnils opened the pull request:
Do not download if an external path is already provided.
Force download (immediately call download script) when using recursive_find.

This is fixing issue: https://github.com/tensorflow/tensorflow/issues/44261

mansnils committed:
TFLu: Update builds with CMSIS-NN

- Do not download CMSIS-NN if not needed.
- Do not patch CMSIS-NN if not needed.

Change-Id: Ia5d5d5714eeb0afdcdc2acb44e63ba9de6b65047

felix-johnny commented on tensorflow/lite/micro/kernels/cmsis-nn/conv.cc:
```diff
@@ -13,10 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "arm_nnfunctions.h"
```
@advaitjain CMSIS-NN provides arm_nnfunctions.h as the sole interface API file to its users. For the glue functions here, I would say that it is just the API's that are of interest than how the arguments are constructed. Else, it becomes a double addition(not that it is an issue) which is superfluous.
 
 I would suggest to go with just arm_nnfunctions.h as the include file here.

advaitjain replied:
Good to know. I'll let you guys decide since it is your API. If you'd like to go with only arm_nnfunctions.h (with the full path), that is fine with me.

gbaned commented:
@mansnils  Can you please check @advaitjain's comments and resolve conflicts?. Thanks!

mansnils commented on tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:
```diff
@@ -1,133 +1,54 @@
+    $(call recursive_find,$(CMSIS_PATH)/CMSIS/NN/Source,*.c)
+  THIRD_PARTY_CC_HDRS += \
+    $(call recursive_find,$(CMSIS_PATH)/CMSIS/NN/Include,*.h)
+  THIRD_PARTY_CC_HDRS += \
+    $(call recursive_find,$(CMSIS_PATH)/CMSIS/DSP/Include,*.h)
+  THIRD_PARTY_CC_HDRS += \
+    $(call recursive_find,$(CMSIS_PATH)/CMSIS/Core/Include,*.h)
```
Ok, so no wildcards in this PR.

mansnils replied:
If we completely remove CMSIS patching would you be ok to use wildcards then?

mansnils commented on tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:
```diff
@@ -1,133 +1,54 @@
-      $(CMSIS_PATH)/CMSIS/NN/Source/SoftmaxFunctions/arm_softmax_with_batch_q7.c \
-      $(CMSIS_PATH)/CMSIS/NN/Source/SVDFunctions/arm_svdf_s8.c \
-      $(CMSIS_PATH)/CMSIS/NN/Source/ReshapeFunctions/arm_reshape_s8.c
+  # Download and optionally patch CMSIS repository.
+  # This needs to be done before recursive_find below as it won't work with the standard
+  # way of downloading by generating a target (call add_third_party_download).
+  # So instead use the shell function, which will immediately download it.
+  # Optional argument is: patch_cmsis
+  DOWNLOAD_SCRIPT := $(MAKEFILE_DIR)/download_and_extract.sh
+  DOWNLOAD_CMSIS = $(or $(shell $(DOWNLOAD_SCRIPT) $(CMSIS_URL) $(CMSIS_MD5) $(CMSIS_PATH) $(1) >&2 && echo SUCCESS), $(error $(DOWNLOAD_SCRIPT) failed))
+  DOWNLOAD_CMSIS_IF_NEEDED = $(if $(shell test -d $(CMSIS_PATH) || echo NEED_DOWNLOAD), $(call DOWNLOAD_CMSIS, $(1)))
```
The shell function is needed when using wildcards because it will download before find_recursive is evaluated, as opposed to the eval function. However since we remove wildcards, we can use eval.

advaitjain replied:
sounds good.

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/add.cc:
```diff
@@ -13,9 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "arm_nnfunctions.h"
```
That wouldn't work when supplying an external CMSIS_PATH on command line.

When having
#include "arm_nnfunctions.h"
it can be found because we have
  INCLUDES += \
    -I$(CMSIS_PATH)/CMSIS/Core/Include \
    -I$(CMSIS_PATH)/CMSIS/DSP/Include \
    -I$(CMSIS_PATH)/CMSIS/NN/Include

When having 
#include "cmsis/CMSIS/NN/Include/arm_nnfunctions.h"
it can be found because we have this in Makefile
INCLUDES := \
-I. \
-I$(MAKEFILE_DIR)/downloads/ \

However with an external CMSIS_PATH there is no 'cmsis' in lower case in the path.

I propose to keep it as just
#include "arm_nnfunctions.h"
If that is OK?

advaitjain replied:
again, thinking some more, this might get in the way of Arduino.

Let me fix a few things for the Arduino before any updates on this PR.

mansnils replied:
Ok

mansnils commented on tensorflow/lite/micro/tools/make/Makefile:
```diff
@@ -362,6 +362,9 @@ $(eval $(call add_third_party_download,$(RUY_URL),$(RUY_MD5),ruy,))
 $(eval $(call add_third_party_download,$(PERSON_MODEL_INT8_URL),$(PERSON_MODEL_INT8_MD5),person_model_int8,))
 
+TARGETS_USING_QUALIFIED_CMSIS_NN_INCLUDE_PATH := \
+arduino
```
Ok. Just a note as discussed, I don't think CMSIS patching is needed for any target, however keeping those changes in a separate PR.

advaitjain replied:
My bad, I misspoke. I thought ESP32 was based on Cortex-M but it is not.

So it looks like we only need to test Arduino and mbed. If CMSIS patching is not needed for either of these two then I'm ok with removing patching completely in one go to keep things simpler.

mansnils replied:
Ok, let's remove CMSIS patching completely then

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/depthwise_conv.cc:
```diff
@@ -13,9 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "arm_nnfunctions.h"
```
See previous comments.

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/fully_connected.cc:
```diff
@@ -13,9 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "arm_nnfunctions.h"
```
See previous comments.

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/mul.cc:
```diff
@@ -13,9 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "arm_nnfunctions.h"
```
See previous comments.

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc:
```diff
@@ -13,9 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#include "arm_nnfunctions.h"
```
See previous comments.

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/svdf.cc:
```diff
@@ -16,8 +16,7 @@ limitations under the License.
 
-#include "cmsis/CMSIS/NN/Include/arm_nn_types.h"
-#include "cmsis/CMSIS/NN/Include/arm_nnfunctions.h"
+#include "arm_nnfunctions.h"
```
See previous comments.

mansnils committed:
TFLu: Update from review comments

Also update related CMSIS download for Ethos-U.

mansnils committed:
Merge remote-tracking branch origin/upstream/master

advaitjain reviewed (commented):
Just highlighting here as well that I misspoke about other targets relying on CMSIS. If Arduino and mbed are ok without patching then we can remove it completely with this PR to keep things simple.

Other than that your proposed changes look good to me.

advaitjain commented:
> Just highlighting here as well that I misspoke about other targets relying on CMSIS. If Arduino and mbed are ok without patching then we can remove it completely with this PR to keep things simple.
> 
> Other than that your proposed changes look good to me.

Did some more digging and chatted with @petewarden and we will need to patch for the Arduino even though the `test_arduino.sh` script passes.

Let's stick to the plan of continuing to patch for the Arduino and nothing else. Separately, I'm going to change the script so that it will catch this issue.

advaitjain reviewed (commented):
Apologies for changing my mind multiple times here. Let's pause on this PR while I fix the Arduino test scripts to get us to a point where we have more confidence that we won't break the Arduino.

advaitjain commented:
@mansnils: I created https://github.com/tensorflow/tensorflow/pull/44406 to first remove the patching of the downloaded CMSIS (and partially document why the Arduino still works).

Next I think we might be able to remove the leading cmsis in the include paths and then we should be good in terms of supporting a non-downloaded CMSIS. I'll create one more PR and send it your way tomorrow.

mansnils commented:
@advaitjain  Cool, CMSIS patching is now removed. I wait for your next PR then to remove the leading cmsis in the include path. After that I propose to go back to using wildcards as it makes maintenance easier (less merge conflicts). I can make a PR for that or update this.

mansnils committed:
Merge remote-tracking branch origin/upstream/master

mansnils committed:
TFLu: Do not download CMSIS if external path is provided

If an external CMSIS_PATH is provided CMSIS should not be downloaded.
Use recursive find when downloading CMSIS-NN.
Add download_third_party_download to be used with recursive_find.

advaitjain commented:
> @advaitjain Cool, CMSIS patching is now removed. I wait for your next PR then to remove the leading cmsis in the include path. After that I propose to go back to using wildcards as it makes maintenance easier (less merge conflicts). I can make a PR for that or update this.

Wildcards for the CMSIS/NN directory sounds ok to me. Let's keep the full list for files in Core and DSP.

advaitjain commented on tensorflow/lite/micro/tools/make/helper_functions.inc:
```diff
@@ -511,13 +512,27 @@ define add_third_party_download
 THIRD_PARTY_DOWNLOADS += $(1)!$(2)!tensorflow/lite/micro/tools/make/downloads/$(3)!$(4)!$(5)
 endef
 
+# Helper function
+call_download_script = $(or $(shell $(DOWNLOAD_SCRIPT) $(1) $(2) $(MAKEFILE_DIR)/downloads/$(3) $(4) $(5) >&2 && echo SUCCESS), $(error $(DOWNLOAD_SCRIPT) failed))
+
+# Same as add_third_party_download except it immediately calls download script,
+# which is needed when using recursive_find on downloaded third party.
+# 1 - URL to download archive file from (can be .zip, .tgz, or .bz).
+# 2 - MD5 sum of archive, to check integrity. Use md5sum tool to generate.
+# 3 - Folder name to unpack library into, inside tf/l/x/m/t/downloads root.
+# 4 - Optional patching action, must match clause in download_and_extract.sh.
+# 5 - Optional patching action parameter
+define download_third_party_download
+download_third_party = $(if $(shell test -d $(MAKEFILE_DIR)/downloads/$(3) || echo NEED_DOWNLOAD), $(call call_download_script,$(1),$(2),$(3),$(4),$(5)))
+endef
```
since we aren't adding any rules via make, can we pull this logic out of the makefile and into download_and_extract.sh? I think that would be clearer.

I'm thinking that we add the logic that checks if the downloaded folder already exists to download_and_exract and also add an additional optional parameter to force the download.

And then directly call `download_and_extract` from cmsis_nn.inc and ethosu.inc?

mansnils replied:
We can at least move the logic to only download if not already downloaded to the download script.

advaitjain commented on tensorflow/lite/micro/tools/make/ext_libs/cmsis_nn.inc:
```diff
@@ -5,118 +5,39 @@ ifneq ($(filter cmsis-nn,$(ALL_TAGS)),)
-      $(CMSIS_PATH)/CMSIS/NN/Source/ReshapeFunctions/arm_reshape_s8.c
+    CMSIS_DEFAULT_DOWNLOAD_PATH := $(MAKEFILE_DIR)/downloads/cmsis
+    CMSIS_PATH := $(CMSIS_DEFAULT_DOWNLOAD_PATH)
+    ifeq ($(CMSIS_PATH), $(CMSIS_DEFAULT_DOWNLOAD_PATH))
+      $(eval $(call download_third_party_download,$(CMSIS_URL),$(CMSIS_MD5),cmsis,))
+    endif
```
nit: please add a comment here (and in ethosu.inc):
```make
# We force a download during the first phase of make so that the files exist prior to the
# call to recursive_find below. add_third_party_download prevents the use of wildcards
# and recursive_find in selecting which files to add to THIRD_PARTY_SRCS

mansnils replied:
Ok

advaitjain reviewed (changes requested):
Thanks for your patience with this pull request.

One more round of requested changes -- I'm trying to keep logic that we don't need out of the Makefiles because that is the direction that we are headed for the download logic.

And please update the PR description too.

mansnils committed:
TFLu: update download script logic from review comments
